llvm/tools/repo-fragments/RepoFragments.cpp

```diff
@@ -31,69 +89,71 @@ cl::opt<bool>
     errs() << "Error: '" << TicketPath << "' ("
            << DigestOrError.getError().message() << ")\n";
     return EXIT_FAILURE;
   }
 
-  pstore::database Db(getRepoPath(), pstore::database::access_mode::read_only);
+  pstore::database Db{getRepoPath(RepoPath),
+                      pstore::database::access_mode::read_only};
   const auto CompilationIndex =
       pstore::index::get_index<pstore::trailer::indices::compilation>(Db);
   if (!CompilationIndex) {
-    errs() << "Error: compilation index was not found. \n";
+    errs() << "Error: compilation index was not found.\n";
     return EXIT_FAILURE;
   }
 
-  pstore::index::digest const &Digest = DigestOrError.get();
-  auto CompilationPos = CompilationIndex->find(Db, Digest);
+  const auto CompilationPos = CompilationIndex->find(Db, *DigestOrError);
   if (CompilationPos == CompilationIndex->end(Db)) {
-    errs() << "Error: compilation " << Digest << " was not found.\n";
+    errs() << "Error: compilation " << *DigestOrError << " was not found.\n";
     return EXIT_FAILURE;
   }
 
-  auto Compilation =
-      pstore::repo::compilation::load(Db, CompilationPos->second);
-
-  auto *Sep = "";
-  for (auto const &CM : *Compilation) {
-    assert(CM.name != pstore::typed_address<pstore::indirect_string>::null());
-    const auto MemberName =
-        pstore::indirect_string::read(Db, CM.name).to_string();
-    if (Names.empty() ||
-        (std::find(Names.begin(), Names.end(), MemberName) != Names.end())) {
-      outs() << Sep;
-      Sep = (UseComma ? "," : "\n");
-      if (!DigestOnly)
-        outs() << MemberName << ": ";
-      outs() << CM.digest;
-    }
+  // Copy CommandLineNames to a set. This both eliminates any duplicates and
+  // gives us a fast find() function which can quickly determine whether a name
+  // was requested by the user. (Unfortunately, DenseSet<> doesn't support
+  // std::inserter(), so this is a hand-coded loop rather than a call to
+  // std::copy().)
+  DenseSet<StringRef> CommandLineNames{clamp_to_unsigned(Names.size())};
+  for (const std::string &N : Names) {
+    CommandLineNames.insert(N);
   }
-  outs() << "\n";
 
+  // Is 'Name' present in the CommandLineNames set? If CommandLineNames is
+  // empty, then we consider this to be a set containing all names and always
+  // return true.
+  const auto IsRequestedName = [&CommandLineNames](const StringRef &Name) {
+    return CommandLineNames.empty() ||
+           CommandLineNames.find(Name) != std::end(CommandLineNames);
+  };
+
+  const auto *Sep = "";
+  for (auto const &CM :
+       *pstore::repo::compilation::load(Db, CompilationPos->second)) {
+    // Convert the pstore indirect-string to a StringRef. We don't allocate any
+    // unnecessary memory in these steps.
+    pstore::shared_sstring_view Owner;
+    const auto MemberNameView =
+        pstore::indirect_string::read(Db, CM.name).as_db_string_view(&Owner);
+    StringRef MemberNameRef{MemberNameView.data(), MemberNameView.length()};
```

Comment:
This sequence avoid creating a std::string instance which will require both a memory allocation and a copy of the string body. pstore has `sstring_view` which is basically the same as LLVM’s `StringRef` (and C++ 17’s `string_view` from which it takes its name) except that it uses a shared_ptr<> rather than a raw pointer. This builds on pstore’s use of `shared_ptr<>` to avoid copying.